File: CppSolution/ValidateBinarySearchTree.cpp
#include <iostream>
#include "Utility.h"
using namespace std;

class Solution
{
private:
    bool helper(TreeNode *root, int &pre)
    {
        if (root == NULL)
            return true;
        bool left = helper(root->left, pre);
        if (root->val <= pre)
            return false;
        pre = root->val;
        return left && helper(root->right, pre);
    }
public:
    bool isValidBST(TreeNode *root)
    {
        int pre = INT_MIN;
        return helper(root, pre);
    }
};

int main()
{
    return 0;
}

File: JavaSolution/src/Microsoft/ValidateBinarySearchTree.java
package Microsoft;

/**
 * Created by xyunpeng on 2/20/16.
 */
public class ValidateBinarySearchTree {
    public boolean isValidBST(TreeNode root) {
        return helper(root, null, null);
    }

    private boolean helper(TreeNode root, Integer min, Integer max) {
        if (root == null) {
            return true;
        }

        if (min != null && min >= root.val) {
            return false;
        }

        if (max != null && max <= root.val) {
            return false;
        }

        return helper(root.left, min, root.val) && helper(root.right, root.val, max);
    }
}
